File: rad/state/state.go
// Package state manages the accepted schema state stored in a Rad project.
// The database remains authoritative; this package writes only server-returned
// accepted schemas and refuses to overwrite divergent history.
package state

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
	"github.com/Southclaws/rad/rad/engine/02_catalog/schema"
	"github.com/Southclaws/rad/rad/protocol"
)

const FormatVersion = 1

type Lock struct {
	FormatVersion int    `json:"format_version"`
	SchemaVersion uint64 `json:"schema_version"`
	SchemaHash    string `json:"schema_hash"`
	Snapshot      string `json:"snapshot"`
}

type Accepted struct {
	Lock   Lock
	Source []byte
	Schema model.Schema
}

type DivergenceError struct {
	Path     string
	Expected string
	Actual   string
}

func (e *DivergenceError) Error() string {
	return fmt.Sprintf("schema history diverged at %s: expected %s, found %s", e.Path, e.Expected, e.Actual)
}

type Store struct {
	Root string
}

func New(root string) Store { return Store{Root: root} }

func SnapshotName(version uint64) string {
	return fmt.Sprintf("%08d.rad.schema.yaml", version)
}

func (s Store) LockPath() string { return filepath.Join(s.Root, "schema.lock.json") }

func (s Store) SnapshotPath(version uint64) string {
	return filepath.Join(s.Root, "changelog", SnapshotName(version))
}

func (s Store) Load() (Accepted, error) {
	raw, err := os.ReadFile(s.LockPath())
	if err != nil {
		return Accepted{}, err
	}
	var lock Lock
	if err := json.Unmarshal(raw, &lock); err != nil {
		return Accepted{}, fmt.Errorf("%s: %w", s.LockPath(), err)
	}
	if lock.FormatVersion != FormatVersion {
		return Accepted{}, fmt.Errorf("%s: unsupported format_version %d", s.LockPath(), lock.FormatVersion)
	}
	wantSnapshot := filepath.ToSlash(filepath.Join("changelog", SnapshotName(lock.SchemaVersion)))
	if lock.Snapshot != wantSnapshot {
		return Accepted{}, fmt.Errorf("%s: snapshot is %q, want %q", s.LockPath(), lock.Snapshot, wantSnapshot)
	}
	snapshotPath, err := s.resolveSnapshot(lock.Snapshot)
	if err != nil {
		return Accepted{}, err
	}
	source, err := os.ReadFile(snapshotPath)
	if err != nil {
		return Accepted{}, err
	}
	canonical, hash, err := ParseSnapshot(snapshotPath, source)
	if err != nil {
		return Accepted{}, err
	}
	if hash != lock.SchemaHash {
		return Accepted{}, &DivergenceError{Path: snapshotPath, Expected: lock.SchemaHash, Actual: hash}
	}
	return Accepted{Lock: lock, Source: source, Schema: canonical}, nil
}

func (s Store) WriteAccepted(server protocol.SchemaState) (Accepted, error) {
	accepted, err := s.WriteSnapshot(server)
	if err != nil {
		return Accepted{}, err
	}
	if err := s.WriteLock(accepted.Lock); err != nil {
		return Accepted{}, err
	}
	return accepted, nil
}

func (s Store) WriteSnapshot(server protocol.SchemaState) (Accepted, error) {
	canonical, err := canonicalSchema(server.Schema)
	if err != nil {
		return Accepted{}, err
	}
	hash, err := canonical.Hash()
	if err != nil {
		return Accepted{}, err
	}
	if hash != server.SchemaHash {
		return Accepted{}, &DivergenceError{
			Path: "server accepted schema", Expected: server.SchemaHash, Actual: hash,
		}
	}
	source, err := schema.Render(canonical)
	if err != nil {
		return Accepted{}, err
	}
	snapshotPath := s.SnapshotPath(server.SchemaVersion)
	if err := os.MkdirAll(filepath.Dir(snapshotPath), 0o755); err != nil {
		return Accepted{}, err
	}
	if existing, err := os.ReadFile(snapshotPath); err == nil {
		_, existingHash, parseErr := ParseSnapshot(snapshotPath, existing)
		if parseErr != nil {
			return Accepted{}, parseErr
		}
		if existingHash != server.SchemaHash {
			return Accepted{}, &DivergenceError{
				Path: snapshotPath, Expected: server.SchemaHash, Actual: existingHash,
			}
		}
		source = existing
	} else if !errors.Is(err, os.ErrNotExist) {
		return Accepted{}, err
	} else if err := atomicCreate(snapshotPath, source, 0o644); err != nil {
		if !errors.Is(err, os.ErrExist) {
			return Accepted{}, err
		}
		existing, readErr := os.ReadFile(snapshotPath)
		if readErr != nil {
			return Accepted{}, readErr
		}
		_, existingHash, parseErr := ParseSnapshot(snapshotPath, existing)
		if parseErr != nil {
			return Accepted{}, parseErr
		}
		if existingHash != server.SchemaHash {
			return Accepted{}, &DivergenceError{
				Path: snapshotPath, Expected: server.SchemaHash, Actual: existingHash,
			}
		}
		source = existing
	}

	lock := Lock{
		FormatVersion: FormatVersion,
		SchemaVersion: server.SchemaVersion,
		SchemaHash:    server.SchemaHash,
		Snapshot:      filepath.ToSlash(filepath.Join("changelog", SnapshotName(server.SchemaVersion))),
	}
	return Accepted{Lock: lock, Source: source, Schema: canonical}, nil
}

func (s Store) WriteLock(lock Lock) error {
	lockSource, err := json.MarshalIndent(lock, "", "  ")
	if err != nil {
		return err
	}
	lockSource = append(lockSource, '\n')
	if err := atomicWrite(s.LockPath(), lockSource, 0o644); err != nil {
		return err
	}
	return nil
}

func (s Store) WriteDesired(filename string, source []byte) error {
	if _, _, err := ParseSnapshot(filename, source); err != nil {
		return err
	}
	return atomicWrite(filename, source, 0o644)
}

func (s Store) BackupDesired(filename string, now time.Time) (string, error) {
	source, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	directory := filepath.Join(s.Root, "backups")
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return "", err
	}
	name := now.UTC().Format("20060102T150405Z") + ".rad.schema.yaml"
	path := filepath.Join(directory, name)
	if _, err := os.Stat(path); err == nil {
		return "", fmt.Errorf("backup already exists: %s", path)
	} else if !errors.Is(err, os.ErrNotExist) {
		return "", err
	}
	if err := atomicCreate(path, source, 0o644); err != nil {
		if errors.Is(err, os.ErrExist) {
			return "", fmt.Errorf("backup already exists: %s", path)
		}
		return "", err
	}
	return path, nil
}

func ParseSnapshot(filename string, source []byte) (model.Schema, string, error) {
	parsed, err := schema.Parse(filename, source)
	if err != nil {
		return model.Schema{}, "", err
	}
	canonical := parsed.Canonical()
	hash, err := canonical.Hash()
	return canonical, hash, err
}

func MatchesAccepted(filename string, source []byte, accepted Accepted) (bool, string, error) {
	_, hash, err := ParseSnapshot(filename, source)
	if err != nil {
		return false, "", err
	}
	return hash == accepted.Lock.SchemaHash, hash, nil
}

func (s Store) resolveSnapshot(relative string) (string, error) {
	clean := filepath.Clean(filepath.FromSlash(relative))
	if clean == "." || filepath.IsAbs(clean) || clean == ".." || len(clean) >= 3 && clean[:3] == ".."+string(filepath.Separator) {
		return "", fmt.Errorf("%s: snapshot path escapes rad.state: %q", s.LockPath(), relative)
	}
	return filepath.Join(s.Root, clean), nil
}

func canonicalSchema(document protocol.SchemaDocument) (model.Schema, error) {
	definitions := make([]model.TableDef, len(document.Tables))
	for i, table := range document.Tables {
		definition := model.TableDef{
			ID: model.SchemaID(table.ID), Name: table.Name, PrimaryKey: table.PrimaryKey,
		}
		for _, column := range table.Columns {
			defaultValue, err := canonicalDefault(column)
			if err != nil {
				return model.Schema{}, err
			}
			definition.Columns = append(definition.Columns, model.ColumnDef{
				ID: model.SchemaID(column.ID), Name: column.Name, Type: model.Type(column.Type),
				Nullable: column.Nullable, Format: column.Format, Default: defaultValue,
			})
		}
		for _, index := range table.Indexes {
			definition.Indexes = append(definition.Indexes, model.IndexDef{
				Name: index.Name, Columns: index.Columns, Unique: index.Unique,
			})
		}
		for _, foreignKey := range table.ForeignKeys {
			definition.ForeignKeys = append(definition.ForeignKeys, model.ForeignKeyDef{
				Name: foreignKey.Name, Columns: foreignKey.Columns,
				RefTable: foreignKey.RefTable, RefColumns: foreignKey.RefColumns,
			})
		}
		definitions[i] = definition
	}
	return model.SchemaFromDefinitions(definitions), nil
}

func canonicalDefault(column protocol.ColumnDef) (*model.Default, error) {
	if column.Default == nil {
		return nil, nil
	}
	if column.Default.Func != "" {
		return &model.Default{Func: model.DefaultFunc(column.Default.Func)}, nil
	}
	value := column.Default.Value
	switch model.Type(column.Type) {
	case model.TypeText:
		text, ok := value.(string)
		if !ok {
			return nil, fmt.Errorf("schema state: column %s default is %T, want string", column.Name, value)
		}
		return &model.Default{Text: text}, nil
	case model.TypeBool:
		boolean, ok := value.(bool)
		if !ok {
			return nil, fmt.Errorf("schema state: column %s default is %T, want bool", column.Name, value)
		}
		return &model.Default{Bool: boolean}, nil
	case model.TypeInt64:
		var number json.Number
		switch typed := value.(type) {
		case json.Number:
			number = typed
		case float64:
			number = json.Number(fmt.Sprint(typed))
		default:
			number = json.Number(fmt.Sprint(typed))
		}
		integer, err := number.Int64()
		if err != nil {
			return nil, fmt.Errorf("schema state: column %s integer default: %w", column.Name, err)
		}
		return &model.Default{Int64: integer}, nil
	case model.TypeFloat64:
		number := json.Number(fmt.Sprint(value))
		floating, err := number.Float64()
		if err != nil {
			return nil, fmt.Errorf("schema state: column %s float default: %w", column.Name, err)
		}
		return &model.Default{Float64: floating}, nil
	default:
		return nil, fmt.Errorf("schema state: column %s has unsupported type %q", column.Name, column.Type)
	}
}

File: src/cli/state.rs
use std::path::{Component, Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::engine::catalog::identity::SchemaId;
use crate::engine::catalog::model::{
    ColumnDef, DefaultFunction, DefaultValue, ForeignKeyDef, IndexDef, ScalarType, Schema, TableDef,
};
use crate::http::generated::types as wire;
use crate::process::Result;

const FORMAT_VERSION: u32 = 1;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub(super) struct Lock {
    pub format_version: u32,
    pub schema_version: u64,
    pub schema_hash: String,
    pub snapshot: String,
}

#[derive(Clone, Debug)]
pub(super) struct Accepted {
    pub lock: Lock,
    pub source: Vec<u8>,
}

pub(super) struct StateStore {
    root: PathBuf,
}

impl StateStore {
    pub(super) fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    pub(super) fn lock_path(&self) -> PathBuf {
        self.root.join("schema.lock.json")
    }

    pub(super) fn snapshot_path(&self, version: u64) -> PathBuf {
        self.root.join("changelog").join(snapshot_name(version))
    }

    pub(super) fn load(&self) -> Result<Accepted> {
        let lock_path = self.lock_path();
        let lock: Lock = serde_json::from_slice(&read_file(&lock_path)?)
            .map_err(|error| format!("{}: {error}", lock_path.display()))?;
        if lock.format_version != FORMAT_VERSION {
            return Err(format!(
                "{}: unsupported format_version {}",
                lock_path.display(),
                lock.format_version
            )
            .into());
        }
        let expected = format!("changelog/{}", snapshot_name(lock.schema_version));
        if lock.snapshot != expected {
            return Err(format!(
                "{}: invalid snapshot {:?}; expected {:?}",
                lock_path.display(),
                lock.snapshot,
                expected
            )
            .into());
        }
        let snapshot = safe_snapshot_path(&self.root, &lock.snapshot)?;
        let source = read_file(&snapshot)?;
        let hash = parse_snapshot(&snapshot, &source)?.hash()?;
        if hash != lock.schema_hash {
            return Err(divergence(&snapshot, &lock.schema_hash, &hash));
        }
        Ok(Accepted { lock, source })
    }

    pub(super) fn write_accepted(&self, state: wire::SchemaState) -> Result<Accepted> {
        let accepted = self.write_snapshot(state)?;
        self.write_lock(&accepted.lock)?;
        Ok(accepted)
    }

    pub(super) fn write_snapshot(&self, state: wire::SchemaState) -> Result<Accepted> {
        let version = u64::try_from(state.schema_version)
            .map_err(|_| format!("invalid server schema version {}", state.schema_version))?;
        let schema = canonical_schema(state.schema)?;
        let hash = schema.hash()?;
        if hash != state.schema_hash {
            return Err(divergence(
                Path::new("server accepted schema"),
                &state.schema_hash,
                &hash,
            ));
        }
        let mut source = crate::engine::catalog::schema::render(&schema)?;
        let snapshot = self.snapshot_path(version);
        if snapshot.exists() {
            let existing = read_file(&snapshot)?;
            let existing_hash = parse_snapshot(&snapshot, &existing)?.hash()?;
            if existing_hash != hash {
                return Err(divergence(&snapshot, &hash, &existing_hash));
            }
            source = existing;
        } else {
            if let Some(parent) = snapshot.parent() {
                std::fs::create_dir_all(parent)?;
            }
            match atomic_create(&snapshot, &source) {
                Ok(()) => {}
                Err(error) if error.kind() == std::io::ErrorKind::AlreadyExists => {
                    let existing = read_file(&snapshot)?;
                    let existing_hash = parse_snapshot(&snapshot, &existing)?.hash()?;
                    if existing_hash != hash {
                        return Err(divergence(&snapshot, &hash, &existing_hash));
                    }
                    source = existing;
                }
                Err(error) => return Err(error.into()),
            }
        }
        Ok(Accepted {
            lock: Lock {
                format_version: FORMAT_VERSION,
                schema_version: version,
                schema_hash: hash,
                snapshot: format!("changelog/{}", snapshot_name(version)),
            },
            source,
        })
    }

    pub(super) fn write_lock(&self, lock: &Lock) -> Result {
        let mut source = serde_json::to_vec_pretty(lock)?;
        source.push(b'\n');
        atomic_write(&self.lock_path(), &source)
    }

    pub(super) fn write_desired(path: &Path, source: &[u8]) -> Result {
        parse_snapshot(path, source)?;
        atomic_write(path, source)
    }

    pub(super) fn backup_desired(&self, path: &Path, now: DateTime<Utc>) -> Result<PathBuf> {
        let source = read_file(path)?;
        let backup = self
            .root
            .join("backups")
            .join(format!("{}.rad.schema.yaml", now.format("%Y%m%dT%H%M%SZ")));
        if let Some(parent) = backup.parent() {
            std::fs::create_dir_all(parent)?;
        }
        atomic_create(&backup, &source)?;
        Ok(backup)
    }
}

pub(super) fn matches_accepted(path: &Path, source: &[u8], accepted: &Accepted) -> Result<bool> {
    Ok(parse_snapshot(path, source)?.hash()? == accepted.lock.schema_hash)
}

fn snapshot_name(version: u64) -> String {
    format!("{version:08}.rad.schema.yaml")
}

fn parse_snapshot(path: &Path, source: &[u8]) -> Result<Schema> {
    Ok(crate::engine::catalog::schema::parse(&path.display().to_string(), source)?.canonical())
}

fn read_file(path: &Path) -> std::io::Result<Vec<u8>> {
    std::fs::read(path)
        .map_err(|error| std::io::Error::new(error.kind(), format!("{}: {error}", path.display())))
}

fn canonical_schema(document: wire::SchemaDocument) -> Result<Schema> {
    let mut tables = Vec::with_capacity(document.tables.len());
    for table in document.tables {
        let id = schema_id(table.id, "table")?;
        let mut columns = Vec::with_capacity(table.columns.len());
        for column in table.columns {
            let scalar_type = match column.r#type.as_str() {
                "text" => ScalarType::Text,
                "int64" => ScalarType::Int64,
                "float64" => ScalarType::Float64,
                "bool" => ScalarType::Bool,
                value => return Err(format!("unsupported column type {value:?}").into()),
            };
            columns.push(ColumnDef {
                id: schema_id(column.id, "column")?,
                name: column.name.clone(),
                scalar_type,
                nullable: column.nullable.unwrap_or(false),
                format: column.format.unwrap_or_default(),
                default: decode_default(&column.name, scalar_type, column.default)?,
            });
        }
        tables.push(TableDef {
            id,
            name: table.name,
            columns,
            primary_key: table.primary_key,
            indexes: table
                .indexes
                .unwrap_or_default()
                .into_iter()
                .map(|index| IndexDef {
                    name: index.name,
                    columns: index.columns,
                    unique: index.unique.unwrap_or(false),
                })
                .collect(),
            foreign_keys: table
                .foreign_keys
                .unwrap_or_default()
                .into_iter()
                .map(|foreign_key| ForeignKeyDef {
                    name: foreign_key.name,
                    columns: foreign_key.columns,
                    ref_table: foreign_key.ref_table,
                    ref_columns: foreign_key.ref_columns,
                })
                .collect(),
        });
    }
    Ok(Schema::from_definitions(tables))
}

fn schema_id(value: Option<i64>, role: &str) -> Result<SchemaId> {
    let value = value.ok_or_else(|| format!("server {role} definition has no schema ID"))?;
    let value = u32::try_from(value)
        .map_err(|_| format!("server {role} schema ID {value} is outside the supported range"))?;
    SchemaId::new(value).map_err(Into::into)
}

fn decode_default(
    column: &str,
    scalar_type: ScalarType,
    default: Option<wire::ColumnDefault>,
) -> Result<Option<DefaultValue>> {
    let Some(default) = default else {
        return Ok(None);
    };
    if let Some(function) = default.func {
        if default.value.is_some() {
            return Err(format!("column {column:?} default sets both func and value").into());
        }
        let function = match function.as_str() {
            "uuid" => DefaultFunction::Uuid,
            "now_ms" => DefaultFunction::NowMs,
            value => return Err(format!("unknown default function {value:?}").into()),
        };
        return Ok(Some(DefaultValue {
            function: Some(function),
            ..DefaultValue::default()
        }));
    }
    let value = default
        .value
        .ok_or_else(|| format!("column {column:?} default has neither func nor value"))?;
    let mut output = DefaultValue::default();
    match scalar_type {
        ScalarType::Text => {
            output.text = value
                .as_str()
                .ok_or_else(|| format!("column {column:?} default expects text"))?
                .into();
        }
        ScalarType::Int64 => {
            output.int64 = value
                .as_i64()
                .ok_or_else(|| format!("column {column:?} default expects int64"))?;
        }
        ScalarType::Float64 => {
            output.float64 = value
                .as_f64()
                .ok_or_else(|| format!("column {column:?} default expects float64"))?;
        }
        ScalarType::Bool => {
            output.bool_value = value
                .as_bool()
                .ok_or_else(|| format!("column {column:?} default expects bool"))?;
        }
    }
    Ok(Some(output))
}

fn safe_snapshot_path(root: &Path, relative: &str) -> Result<PathBuf> {
    let path = Path::new(relative);
    if path.is_absolute()
        || path
            .components()
            .any(|component| matches!(component, Component::ParentDir | Component::RootDir))
    {
        return Err(format!("snapshot path escapes {}: {relative:?}", root.display()).into());
    }
    Ok(root.join(path))
}

fn atomic_create(path: &Path, source: &[u8]) -> std::io::Result<()> {
    use std::io::Write;
    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    let mut file = tempfile::NamedTempFile::new_in(parent)?;
    file.write_all(source)?;
    file.as_file().sync_all()?;
    file.persist_noclobber(path)
        .map(|_| ())
        .map_err(|error| error.error)
}

fn atomic_write(path: &Path, source: &[u8]) -> Result {
    use std::io::Write;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    let mut file = tempfile::NamedTempFile::new_in(parent)?;
    file.write_all(source)?;
    file.as_file().sync_all()?;
    file.persist(path).map_err(|error| error.error)?;
    Ok(())
}

fn divergence(
    path: &Path,
    expected: &str,
    actual: &str,
) -> Box<dyn std::error::Error + Send + Sync> {
    format!(
        "schema history diverged at {}: expected {expected}, found {actual}",
        path.display()
    )
    .into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn missing_state_names_the_expected_lock_file() {
        let directory = tempfile::tempdir().unwrap();
        let store = StateStore::new(directory.path());

        let error = store.load().unwrap_err();

        assert!(
            error.to_string().contains(
                &directory
                    .path()
                    .join("schema.lock.json")
                    .display()
                    .to_string()
            ),
            "{error}"
        );
        assert_eq!(
            error
                .downcast_ref::<std::io::Error>()
                .map(std::io::Error::kind),
            Some(std::io::ErrorKind::NotFound)
        );
    }

    #[test]
    fn invalid_lock_snapshot_names_the_field_and_expected_value() {
        let directory = tempfile::tempdir().unwrap();
        let store = StateStore::new(directory.path());
        std::fs::write(
            store.lock_path(),
            r#"{"format_version":1,"schema_version":0,"schema_hash":"sha256:none","snapshot":""}"#,
        )
        .unwrap();

        let error = store.load().unwrap_err().to_string();

        assert!(error.contains("invalid snapshot \"\""), "{error}");
        assert!(
            error.contains("expected \"changelog/00000000.rad.schema.yaml\""),
            "{error}"
        );
    }

    #[test]
    fn snapshot_paths_cannot_escape_state() {
        assert!(safe_snapshot_path(Path::new("state"), "changelog/one.yaml").is_ok());
        assert!(safe_snapshot_path(Path::new("state"), "../one.yaml").is_err());
        assert!(safe_snapshot_path(Path::new("state"), "/one.yaml").is_err());
    }

    #[test]
    fn accepted_state_round_trips_and_detects_tampering() {
        let directory = tempfile::tempdir().unwrap();
        let store = StateStore::new(directory.path());
        let schema = crate::engine::catalog::schema::parse(
            "fixture",
            b"tables:\n  - id: 1\n    name: users\n    columns:\n      - { id: 1, name: id, type: string, pk: true }\n",
        )
        .unwrap()
        .canonical();
        let state = wire::SchemaState {
            schema_version: 4,
            schema_hash: schema.hash().unwrap(),
            schema: wire::SchemaDocument {
                tables: vec![wire::TableDef {
                    columns: vec![wire::ColumnDef {
                        default: None,
                        format: None,
                        id: Some(1),
                        name: "id".into(),
                        nullable: None,
                        r#type: "text".into(),
                    }],
                    foreign_keys: None,
                    id: Some(1),
                    indexes: None,
                    name: "users".into(),
                    primary_key: vec!["id".into()],
                }],
            },
        };
        store.write_accepted(state).unwrap();
        assert_eq!(store.load().unwrap().lock.schema_version, 4);
        std::fs::write(store.snapshot_path(4), b"tables: []\n").unwrap();
        assert!(store.load().is_err());
    }
}
